### PR TITLE
Enhancement dataframe to csv use zip method

### LINKED ISSUE #39465: BUG: DataFrame to_csv compression with 'zip' use zipfilename as archive name
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

```python
a=pd.DataFrame({'a':[1,2,3,4,5],'b':[4,5,6,7,8]})
a.to_csv('myfile.csv.zip')
```

#### Problem description

when use gzip method, the function works fine, a.to_csv('myfile.csv.gz') will create a gzip file with an archive file named myfile.csv.  
the "zip" method works bad,a.to_csv('myfile.csv.zip') will create a zipfile with an archive file named myfile.csv.zip

### PR DESCRIPTION
**Pull request by CyberQin**
Enhancement dataframe to csv use zip method,now df.to_csv('myfile.csv.zip') works like df.to_csv('myfile.csv.gz'), generate a zip file with a myfile.csv as content

- [ ] closes #39465 
- [ ] tests added / passed,simply test
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

a=pd.DataFrame({'a':[1,2,3,4,5],'b':[4,5,6,7,8]})
a.to_csv(r'z:/myfile.csv.zip')

### CODE REVIEW

**Review by arw2019 — COMMENTED** [review]
looks like this breaks some tests

```
=========================== short test summary info ============================
FAILED pandas/tests/io/formats/test_to_csv.py::TestToCSV::test_to_csv_zip_arguments[None-zip]
FAILED pandas/tests/io/formats/test_to_csv.py::TestToCSV::test_to_csv_zip_arguments[None-infer]
```

File: pandas/io/common.py [hunk 1]
```diff
@@ -726,6 +726,12 @@ def __init__(
         self.archive_name = archive_name
         self.multiple_write_buffer: Optional[Union[StringIO, BytesIO]] = None
 
+        # add csv file name like gz or bz2
+        if archive_name is None and isinstance(file, (os.PathLike, str)):
+            archive_name = os.path.basename(file)
```
Comment (twoertwein) [inline pandas/io/common.py:731]: probably better to replace `os.path.basename` with `stringify_path` (to handle PathLike objects). That might also fix the failing `test_to_csv_compression_encoding_gcs` test.

Comment (HQDragon) [reply pandas/io/common.py:731]: its strange that it passed this test on my win10 laptop.
image

Comment (twoertwein) [reply pandas/io/common.py]: sorry, you probably still need `basename` to avoid having the entire path within the archive

Comment (HQDragon) [reply pandas/io/common.py:731]: start get different ci errors, does my commit affect it?

Comment (twoertwein) [reply pandas/io/common.py:729]: Can you add a small test case. For example, write a compressed CSV file and then open it with zipfile.ZipFile to check the name.

Comment (HQDragon) [reply pandas/io/common.py:779]: add test case and finish pre commit, a new PR is opened #40387

**Pull request closed without merging**